### ISSUE TITLE
Nested calls to the same function produces incorrect code

### ISSUE THREAD

**Issue by GearsDatapacks** [GMDU/zoglin#3]
This code:
```rust
namespace example

fn add(a, b) {
  return a + b
}

fn load() {
  add(1, add(2, 3))
}
```
Generates:
```mcfunction
data modify storage example:add a set value 1
data modify storage example:add a set value 2
data modify storage example:add b set value 3
data modify storage example:add return set value false
function example:add
data modify storage example:add b set from storage example:add return
function example:add
```
Which causes the outer function's `a` parameter to be set to `2`, resulting in an incorrect result.
We should probably handle inner function calls first, but that's more complex than it might sound, because it also needs to take into account calls in other expressions:
```rust
add(1, 2 + add(3, 4))
```

**Comment by GearsDatapacks** [comment]
Another thing to consider:
```rs
add(add(1, 2), add(3, 4))
```
No matter which order we call the inner functions in, one of the arguments to the outer function will be incorrect. We need to store some intermediate variable

**Comment by moxvallix** [comment]
Yeah I guess for functions called within a function argument they need to store their result in to temp variables.

**Comment by GearsDatapacks** [comment]
For now we could just use a naive approach, where any nested calls to the same function would be stored in a temp variable, and then optimise it later. There probably are not many cases we can optimise anyway
